packages/react-router/lib/dom/ssr/routes.tsx

```diff
@@ -412,17 +458,49 @@ export function createClientRoutes(
             }
             return fetchServerAction(singleFetch);
           });
+      } else if (route.clientActionModule) {
+        dataRoute.action = async (
+          args: ActionFunctionArgs,
+          singleFetch?: unknown
+        ) => {
+          invariant(route.clientActionModule);
+          prefetchRouteModuleChunks(route);
+          let { clientAction } = await import(
+            /* @vite-ignore */
+            /* webpackIgnore: true */
+            route.clientActionModule
+          );
+          return clientAction({
+            ...args,
+            async serverAction() {
+              preventInvalidServerHandlerCall("action", route, isSpaMode);
+              return fetchServerAction(singleFetch);
+            },
+          });
+        };
       }
 
       // Load all other modules via route.lazy()
       dataRoute.lazy = async () => {
-        let mod = await loadRouteModuleWithBlockingLinks(
+        if (route.clientLoaderModule || route.clientActionModule) {
+          // If a client loader/action chunk is present, we push the loading of
+          // the main route chunk to the next tick to ensure the downloading of
+          // loader/action chunks takes precedence. This can be seen via their
+          // order in the network tab. Also note that since this is happening
+          // within `route.lazy`, this imperceptible delay only happens on the
+          // first load of this route.
+          await new Promise((resolve) => setTimeout(resolve, 0));
+        }
+
+        let modPromise = loadRouteModuleWithBlockingLinks(
           route,
           routeModulesCache
         );
+        prefetchRouteModuleChunks(route);
+        let mod = await modPromise;
 
         let lazyRoute: Partial<DataRouteObject> = { ...mod };
-        if (mod.clientLoader) {
+        if (mod.clientLoader && !route.clientLoaderModule) {
```

Comment:
These 2 should never be present together right?  If we have a `clientLoaderModule` then the clientLoader should have been stripped from our primary module?

Reply:
Yep, good point. Plus we've already ruled out that the two properties can't co-exist just above the `route.lazy` definition.